WeakAuras/History.lua

```diff
@@ -0,0 +1,88 @@
+--[[
+This file contains all of the interfaces to the history collection. Indexed by uid of the active display.
+
+History:
+  Table containing historical information about display:
+  source: string indicating where this data came from. Possible values:
+    "user": aura was created by the user via the WeakAuras interface
+    "import": aura was imported via WeakAuras.Import (i.e. pasted string or addon channel transmission)
+    "addon": aura was imported via WeakAuras.RegisterDisplay (i.e. an addon requested the addition)
+  data: snapshot of data as it was at last import, or nil if source == "user", and the user has not chosen to save their changes to history
+  addon: if source == "addon", then this string indicates which addon supplied the last import
+  skippedVersions: record of versions that the user has indicated they don't wish to see update notifications for
+  allowUpdates: if false, then WeakAuras will ignore any addon-sourced update requests
+  lastUpdate: UNIX timestamp of the last time the user completed an import for this display
+--]]
+
+local WeakAuras = WeakAuras
+local history -- history db upvalue
+
+function WeakAuras.LoadHistory(historydb)
+  history = historydb
+  if history.clearOldHistory == nil then
+    history.clearOldHistory = 30
+  end
+  if history.clearOldHistory then
+    WeakAuras.ClearOldHistory(history.clearOldHistory)
+  end
+end
+
+function WeakAuras.SetHistory(uid, data, source, addon)
+  if uid and data then
+    history[uid] = history[uid] or {}
+    local hist = history[uid]
+    hist.data = data
+    hist.source = source
+    hist.addon = source == "addon" and (addon or "unknown") or nil
+    hist.skippedVersions = hist.skippedVersions or {}
+    hist.lastUpdate = time()
+  end
+end
```

Comment:
You are overwriting previous data, don't you want to save it with something like `hist.versions[(hist.version or 0)+1] = hist.data` or using timestamp ?

Reply:
No, I don’t. That would bloat the database far more than I’m comfortable doing.

Reply:
Alright, so the idea is to have hist.data always have imported data through transmission in the state it was before any user modification, so you can 3-way compare
1- hist.data (last clean import)
2- data (actual aura with whatever user modification was made on it)
3- new clean update data
Am i right?

Reply:
In a nutshell, yes.